### DISCUSSION TITLE
VPS performance issue ?

### DISCUSSION THREAD

**Discussion by bloor** [Dokploy/dokploy#372] [General]
I noted something weird regarding the additional processing head dokploy/containers add on top of vanilla VPS deploy.

Here are my 2 identical contabo VPS ( 4vCPU / 6GB RAM / 200 GB ).
![image](https://github.com/user-attachments/assets/aa98c417-7712-45a8-876f-6fa99fca00de)

On first one i run the dokploy instance with the shown services in the printscreen.
On second one, I allocated all the 3 webscrapping/puppeteer scripts.

The website is in BETA, not yet released < 10 accounts. So the strain on the website and database is minimal at IDLE. (as seen in the htop graphs)

MY ISSUE:
I tried moving the 3 scripts from 2nd server in the dokploy implementation (creating an Application docker for each script). But running it for ~3 days I had 2 cases of VPS crashing / freezing (100% CPU on all cores) and moreover the IDLE state was constantly 60+% in htop.
As you can see from the printscreen, running them as 2 separate VPS barely reaches 10%+ on IDLE. 

Has anyone experienced something similar with dokploy / raw docker containers ?!
I'm a beginner at docking, but I can't see the performance hit being solely on my configs.
Because, frankly the docker configs for the 3 puppeteer scritps are very simple and light:

```docker
FROM node:20.16.0

RUN apt-get update
RUN apt-get install chromium -y

RUN mkdir -p /app
ADD / /app
WORKDIR /app
RUN npm install
CMD ["node", "index.js"]
```

**Comment by Siumauricio** [comment]
I don't see any issue performances here or memory from dokploy services, dokploy create 4 services when you install:

1. Dokploy (nextjs-application)
![Screenshot 2024-08-22 at 8 11 41 AM](https://github.com/user-attachments/assets/8dad2fc4-fe98-4b9a-a531-f5643d8b3266)

2. Redis database for deployments
![Screenshot 2024-08-22 at 8 13 56 AM](https://github.com/user-attachments/assets/9834ceb8-a9df-4110-ad9b-4b96c0112472)

3. Postgresql
![Screenshot 2024-08-22 at 8 13 47 AM](https://github.com/user-attachments/assets/e378e873-ccc3-47e5-abf1-574e49771f9a)

4. Traefik instance
![Screenshot 2024-08-22 at 8 13 24 AM](https://github.com/user-attachments/assets/a5212bbe-4930-4fdb-9439-3c49115a8c68)

I look in each process and all the values seems very normal the usage is barely to almost 0 so this is not related to dokploy itself

Dokploy doesn't do anything than just building the image and running in to a container, so is basically the same thing you run the commands manually in your server, but dokploy automatize that process

> -----------------------------------------------------------------------

This comparison is not fair btw, you are comparing docker and pm2, pm2 by default is very lightweight , on other hand you need to know docker by default add's some overhead to each container, installing docker comes with pros and cons, also you are using a scrapper just simply try to search related to that library and you will find alot of issues regarding to the cpu usage due to pupeteer uses chrome driver
https://devforth.io/blog/how-to-simply-workaround-ram-leaking-libraries-like-puppeteer-universal-way-to-fix-ram-leaks-once-and-forever/

So I don't think this is a dokploy issue, is more a optimization related to your projects and also if you are running scrappers + multiples websites maybe 6 gb is not enought to handle this

**Comment by bloor** [comment]
I tested 2day back and forth.
I'm asking, to know for sure if I should continue testing and find the issue.
Cause for now I feel defeated. 🫠

There's also a small chance the shared CPUs were under shared-load just those days.
As far as i remember, the 2 applications that were freezing the VPS with cpu overload were Pocketbase (even though it's a minimal install with 3 tables and 10 users) and the Scrapping App (which in isolation barely takes under full running/trigger 1GB)

### RAM could be indeed the issue ###
Guess I'll try again in the weekend, and also maybe increase the resources to 6vCPU / 16GB (since it's literally the same price as now, paying for 2 separate VPS)
Thanks to your insight, i think RAM could be indeed the issue, whenever all the apps are under load , the 5-6GB might be reached which explains the CPUs going crazy trying to heap the memory to SSD.

PS: the project is too good, and i really want to have a 1 place manage all for my project; since it has so many different services (and will have even more).

My favorite features are pocketbase easy install and mysql/pocketbase easy backup activation to S3.

**Reply by Siumauricio** [reply]
Yeah, also some vps providers offer alot of resources but at the end they are don't have truly the resources or atleast the vps are not optimized which can make the vps are really slow, In my case I've used hetzner and the experience is totally differente the performances are really superior

Thank you, I hope you find the real issue on this

**Comment by bloor** [comment]
I narrowed down the issue. 
So I was getting different results, mainly because the VPS is running Ubuntu 22 and the docker containers were defaulted yo Debian 12. But that was not the issue... issue was that they were running different instances of Chromium needed for Puppeteer. Basically for the 3 scripts I had (each in different container, but from same source code) they each took 1 CPU to 100% - even on idle. That meant on old VPS with only 4vCPU, only 1 was free to handle all other 5 containers. Which in turn gave a FULL CPU GRAPH on all cores.

![image](https://github.com/user-attachments/assets/e6fdeda1-2957-40c8-9ca6-e9f11d40bf46)

I actually think the `tutorials` on how to run Chromium on debian, made me install the full browser or something. I fixed with a very light pointed pkg install for chromium and we are in a happy place now.

IDLE: 
![image](https://github.com/user-attachments/assets/f991b7bc-e344-4dab-b6db-5e621fb5de2d)

UNDER LOAD:
![image](https://github.com/user-attachments/assets/65bdc4cf-5ce7-4715-93f4-aeac31c7aebf)
